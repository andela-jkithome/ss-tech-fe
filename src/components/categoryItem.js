import React from 'react';
import PropTypes from 'prop-types';
import { Segment, Grid, Message, Icon } from 'semantic-ui-react';
import ErrorDisplay from './errorDisplay';

function CategoryItem({
  category,
  cat,
  open,
  fetching,
  value,
  fetchfact,
  factError,
}) {
  let panel;
  if (category === cat && open) {
    if (fetching) {
      panel = (
        <Grid.Row>
          <Grid.Column className="jokemessage">
            <Message>
              <Icon name="circle notched" loading />
              <Message.Content>
                <Message.Header>Just one second</Message.Header>
                We are fetching that joke for you.
              </Message.Content>
            </Message>
          </Grid.Column>
        </Grid.Row>
      );
    } else if (factError) {
      panel = (
        <ErrorDisplay
          header={factError}
          text="There was an error fetching the fact. Please try again"
        />
      );
    } else {
      panel = (
        <Grid.Row>
          <Grid.Column>
            <Message
              className="jokemessage"
              header={`${cat.charAt(0).toUpperCase() +
                cat.substr(1)} Chuck Norris Fact`}
              content={value}
            />
          </Grid.Column>
        </Grid.Row>
      );
    }
  } else {
    panel = null;
  }

  return (
    <Grid className="container">
      <Grid.Column columns={1}>
        <Grid.Row>
          <Grid.Column>
            <Segment
              color="blue"
              textAlign="center"
              onClick={() => {
                fetchfact(cat);
              }}
            >
              {cat}
            </Segment>
          </Grid.Column>
        </Grid.Row>
        {panel}
      </Grid.Column>
    </Grid>
  );
}

CategoryItem.propTypes = {
  category: PropTypes.string.isRequired,
  cat: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  value: PropTypes.string,
  fetchfact: PropTypes.func.isRequired,
  factError: PropTypes.string,
};

export default CategoryItem;
