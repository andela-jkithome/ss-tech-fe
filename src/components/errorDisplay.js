import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Message } from 'semantic-ui-react';

function ErrorDisplay({ header, text }) {
  return (
    <Grid.Row>
      <Grid.Column>
        <Message negative className="jokemessage">
          <Message.Header>{header}</Message.Header>
          <p>{text}</p>
        </Message>
      </Grid.Column>
    </Grid.Row>
  );
}

ErrorDisplay.propTypes = {
  header: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
};

export default ErrorDisplay;
