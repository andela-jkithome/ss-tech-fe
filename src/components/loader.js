import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Segment, Grid, Progress } from 'semantic-ui-react';

function Loading({ fetching }) {
  const text = `Click on any one of the categories listed to display a 
  random joke from the category.`;
  return (
    <Fragment>
      {fetching ? (
        <Grid className="container" centered>
          <Grid.Column columns={1}>
            <Grid.Row>
              <Grid.Column>
                <Progress percent={100} indicating />
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column>
                <Segment textAlign="center" color="blue" size="massive">
                  Fetching Categories...
                </Segment>
              </Grid.Column>
            </Grid.Row>
          </Grid.Column>
        </Grid>
      ) : (
        <Grid className="container" centered>
          <Grid.Column columns={1}>
            <Grid.Row>
              <Grid.Column>
                <Segment raised tertiary textAlign="center">
                  {text}
                </Segment>
              </Grid.Column>
            </Grid.Row>
          </Grid.Column>
        </Grid>
      )}
    </Fragment>
  );
}

Loading.propTypes = {
  fetching: PropTypes.bool.isRequired,
};

export default Loading;
