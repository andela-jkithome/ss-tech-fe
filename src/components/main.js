import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Header, Image } from 'semantic-ui-react';
import { fetchCategories } from '../actions/categories';
import { fetchFact } from '../actions/facts';
import CategoryItem from './categoryItem';
import Loader from './loader';
import ErrorDisplay from './errorDisplay';
import logo from '../images/logo.png';

export class Main extends Component {
  constructor() {
    super();
    this.state = {
      open: false,
      category: '',
    };
  }

  componentDidMount() {
    const { fetchCategories: fetchcategories } = this.props;
    fetchcategories('https://api.chucknorris.io/jokes/categories');
  }

  selectCategory = category => {
    const { fetchFact: fetchfact } = this.props;
    const url = `https://api.chucknorris.io/jokes/random?category=${category}`;
    fetchfact(url);
    this.setState({ open: true, category });
  };

  render() {
    const {
      categories: { categories, fetching, error: categoryError },
      facts: {
        fetching: getting,
        fact: { value },
        error: factError,
      },
    } = this.props;
    const { open, category } = this.state;

    return (
      <Fragment>
        <Header as="h3" textAlign="center" color="teal" block>
          Chuck Norris Facts
        </Header>
        <Image className="logo" src={logo} size="medium" centered />
        <Loader fetching={fetching} />
        {categoryError ? (
          <ErrorDisplay
            header={categoryError}
            text="There was an error fetching the categories. Please try again"
          />
        ) : (
          categories.map(cat => (
            <CategoryItem
              key={cat}
              category={category}
              cat={cat}
              open={open}
              fetching={getting}
              fetchfact={this.selectCategory}
              value={value}
              factError={factError}
            />
          ))
        )}
      </Fragment>
    );
  }
}

Main.propTypes = {
  fetchCategories: PropTypes.func.isRequired,
  fetchFact: PropTypes.func.isRequired,
  categories: PropTypes.shape({
    categories: PropTypes.array,
    fetching: PropTypes.bool,
    error: PropTypes.oneOfType([PropTypes.oneOf([null]), PropTypes.string]),
  }).isRequired,
  facts: PropTypes.shape({
    fact: PropTypes.object,
    fetching: PropTypes.bool,
    error: PropTypes.oneOfType([PropTypes.oneOf([null]), PropTypes.string]),
  }).isRequired,
};

function mapStateToProps({ categories, facts }) {
  return {
    categories,
    facts,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchCategories,
      fetchFact,
    },
    dispatch,
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Main);
