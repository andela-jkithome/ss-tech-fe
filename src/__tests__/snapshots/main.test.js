import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Main } from '../../components/main';

configure({ adapter: new Adapter() });

const fact = {
  category: ['science'],
  icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
  id: 'uojt-t8as5ws5h1q-sjslw',
  url: 'https://api.chucknorris.io/jokes/uojt-t8as5ws5h1q-sjslw',
  value: `Nothing can escape the gravity of a black hole, except for Chuck Norris. 
  Chuck Norris eats black holes. They taste like chicken.`,
};

const categories = [
  'explicit',
  'dev',
  'movie',
  'food',
  'celebrity',
  'science',
  'sport',
  'political',
  'religion',
  'animal',
  'history',
  'music',
  'travel',
  'career',
  'money',
  'fashion',
];

const props = {
  categories: {
    fetching: false,
    categories,
    error: null,
  },
  facts: {
    fetching: false,
    fact,
    error: null,
  },
  fetchCategories: jest.fn,
  fetchFact: jest.fn,
};

it('Renders correctly when not fetching', () => {
  const component = shallow(<Main {...props} />);
  expect(toJson(component)).toMatchSnapshot();
});

it('Renders correctly when fetching', () => {
  const component = shallow(
    <Main
      {...Object.assign({}, props, {
        categories: {
          fetching: true,
          categories,
          error: null,
        },
      })}
    />,
  );
  expect(toJson(component)).toMatchSnapshot();
});

it('Renders correctly when there was an error fetching categories', () => {
  const { categories: cats } = props;
  const component = shallow(
    <Main
      {...Object.assign({}, props, {
        categories: {
          ...cats,
          error: 'Error Fetching Categories!',
        },
      })}
    />,
  );
  expect(toJson(component)).toMatchSnapshot();
});
