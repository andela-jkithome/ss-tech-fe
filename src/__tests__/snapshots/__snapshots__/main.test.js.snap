// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly when fetching 1`] = `
<React.Fragment>
  <Header
    as="h3"
    block={true}
    color="teal"
    textAlign="center"
  >
    Chuck Norris Facts
  </Header>
  <Image
    as="img"
    centered={true}
    className="logo"
    size="medium"
    src="logo.png"
    ui={true}
  />
  <Loading
    fetching={true}
  />
  <CategoryItem
    cat="explicit"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="explicit"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="dev"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="dev"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="movie"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="movie"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="food"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="food"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="celebrity"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="celebrity"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="science"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="science"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="sport"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="sport"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="political"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="political"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="religion"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="religion"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="animal"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="animal"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="history"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="history"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="music"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="music"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="travel"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="travel"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="career"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="career"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="money"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="money"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="fashion"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="fashion"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
</React.Fragment>
`;

exports[`Renders correctly when not fetching 1`] = `
<React.Fragment>
  <Header
    as="h3"
    block={true}
    color="teal"
    textAlign="center"
  >
    Chuck Norris Facts
  </Header>
  <Image
    as="img"
    centered={true}
    className="logo"
    size="medium"
    src="logo.png"
    ui={true}
  />
  <Loading
    fetching={false}
  />
  <CategoryItem
    cat="explicit"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="explicit"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="dev"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="dev"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="movie"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="movie"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="food"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="food"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="celebrity"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="celebrity"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="science"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="science"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="sport"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="sport"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="political"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="political"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="religion"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="religion"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="animal"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="animal"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="history"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="history"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="music"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="music"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="travel"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="travel"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="career"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="career"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="money"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="money"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
  <CategoryItem
    cat="fashion"
    category=""
    factError={null}
    fetchfact={[Function]}
    fetching={false}
    key="fashion"
    open={false}
    value="Nothing can escape the gravity of a black hole, except for Chuck Norris. 
    Chuck Norris eats black holes. They taste like chicken."
  />
</React.Fragment>
`;

exports[`Renders correctly when there was an error fetching categories 1`] = `
<React.Fragment>
  <Header
    as="h3"
    block={true}
    color="teal"
    textAlign="center"
  >
    Chuck Norris Facts
  </Header>
  <Image
    as="img"
    centered={true}
    className="logo"
    size="medium"
    src="logo.png"
    ui={true}
  />
  <Loading
    fetching={false}
  />
  <ErrorDisplay
    header="Error Fetching Categories!"
    text="There was an error fetching the categories. Please try again"
  />
</React.Fragment>
`;
